hudi-common/src/main/java/org/apache/hudi/common/table/read/UpdateProcessor.java

```diff
@@ -0,0 +1,128 @@
+   * @param <T> the engine specific record type
+   */
+  class StandardUpdateProcessor<T> implements UpdateProcessor<T> {
+    protected final HoodieReadStats readStats;
+    protected final HoodieReaderContext<T> readerContext;
+    protected final boolean emitDeletes;
+
+    public StandardUpdateProcessor(HoodieReadStats readStats, HoodieReaderContext<T> readerContext,
+                                   boolean emitDeletes) {
+      this.readStats = readStats;
+      this.readerContext = readerContext;
+      this.emitDeletes = emitDeletes;
+    }
+
+    @Override
+    public T processUpdate(String recordKey, T previousRecord, T currentRecord, boolean isDelete) {
+      if (isDelete) {
+        readStats.incrementNumDeletes();
+        if (emitDeletes) {
+          return readerContext.getDeleteRow(currentRecord, recordKey);
+        }
+        return null;
+      } else {
+        if (previousRecord != null) {
+          readStats.incrementNumUpdates();
+        } else {
+          readStats.incrementNumInserts();
+        }
+        return readerContext.seal(currentRecord);
+      }
+    }
+  }
+
+  /**
+   * A processor that wraps the standard update processor and invokes a customizable callback for each update.
+   * @param <T> the engine specific record type
+   */
+  class CallbackProcessor<T> implements UpdateProcessor<T> {
+    private final BaseFileUpdateCallback callback;
+    private final UpdateProcessor<T> delegate;
+    private final HoodieReaderContext<T> readerContext;
+    private final Option<UnaryOperator<T>> outputConverter;
+    private final Schema requestedSchema;
+
+    public CallbackProcessor(BaseFileUpdateCallback callback, UpdateProcessor<T> delegate,
+                             HoodieReaderContext<T> readerContext) {
+      this.callback = callback;
+      this.delegate = delegate;
+      this.readerContext = readerContext;
+      this.outputConverter = readerContext.getSchemaHandler().getOutputConverter();
+      this.requestedSchema = readerContext.getSchemaHandler().getRequestedSchema();
+    }
+
+    @Override
+    public T processUpdate(String recordKey, T previousRecord, T currentRecord, boolean isDelete) {
+      T result = delegate.processUpdate(recordKey, previousRecord, currentRecord, isDelete);
+
+      if (isDelete) {
+        callback.onDelete(recordKey, convertOutput(previousRecord));
+      } else if (previousRecord != null && previousRecord != currentRecord) {
+        callback.onUpdate(recordKey, convertOutput(previousRecord), convertOutput(currentRecord));
+      } else {
+        callback.onInsert(recordKey, convertOutput(currentRecord));
+      }
+      return result;
+    }
+
+    private GenericRecord convertOutput(T record) {
+      T convertedRecord = outputConverter.map(converter -> record == null ? null : converter.apply(record)).orElse(record);
+      return convertedRecord == null ? null : readerContext.convertToAvroRecord(convertedRecord, requestedSchema);
+    }
```

Comment:
Should the `convertOuput` be part of the `CDCCallback` implementation? The logic of `convertOuput` seems unrelated to the processor logic here.  To achieve that, `BaseFileUpdateCallback#onUpdate, #onInsert, #onDelete` should also take instance of `T` instead of `GenericRecord`.

Reply:
True, let me move it